File: solutions/109_lru.go
package main

import "fmt"

// Node 双向链表节点定义
type Node struct {
	key        int // 用于反向获取Cache指针
	value      int
	prev, next *Node
}

type LRUCache struct {
	hashmap  map[int]*Node
	length   int
	capacity int
	head     *Node // 最近最常用key-value指针
	tail     *Node // 最近最不常用key-value指针
}

func Constructor(capacity int) LRUCache {
	rtv := LRUCache{
		hashmap:  make(map[int]*Node),
		capacity: capacity,
		head:     &Node{}, // 虚拟头节点
		tail:     &Node{}, // 虚拟尾节点 便于insert/delete操作
	}
	rtv.head.next = rtv.tail
	rtv.tail.prev = rtv.head
	return rtv
}

func (this *LRUCache) removeNode(node *Node) {
	p, n := node.prev, node.next
	p.next = n
	n.prev = p
	node.next, node.prev = nil, nil
	this.hashmap[node.key] = nil
	this.length--
}

func (this *LRUCache) addNodeToHead(node *Node) {
	p := this.head.next
	this.head.next = node
	node.next = p
	node.prev = this.head
	p.prev = node
	this.length++
}

func (this *LRUCache) Get(key int) int {
	node := this.hashmap[key]
	if node == nil {
		return -1
	}
	// 放到链表头部 表示最近使用过
	this.removeNode(node)
	this.addNodeToHead(node)
	this.hashmap[key] = node
	return node.value
}

func (this *LRUCache) Put(key int, value int) {
	if this.hashmap[key] != nil {
		this.hashmap[key].value = value
		this.Get(key) // Tips: 利用Get放到队头
		return
	}
	if this.length == this.capacity {
		this.removeNode(this.tail.prev) // 先移除最近最少使用的
	}
	// 添加节点
	node := &Node{
		key:   key,
		value: value,
	}
	this.addNodeToHead(node)
	this.hashmap[key] = node
}

func main() {
	//["LRUCache","put","get","put","get","get"]
	//[[1],[2,1],[2],[3,2],[2],[3]]
	obj := Constructor(1)
	obj.Put(2, 1)
	fmt.Println(obj.Get(2))
	obj.Put(3, 2)
	fmt.Println(obj.Get(2))
	fmt.Println(obj.Get(3))
}

File: solutions/__109_lru.cpp
/*
 *      Author: peanutzhen
 *      Created time: 2021/9/3 10:51:17
 */

#include <bits/stdc++.h>

using namespace std;

class LRUCache {
private:
    struct ListNode {
        int key;
        int val;
        ListNode *prev;
        ListNode *next;

        ListNode(int key, int value) : key(key), val(value), prev(nullptr), next(nullptr) {}
    };
    int size;
    int capacity;
    unordered_map<int, ListNode *> hashmap;
    ListNode *head;
    ListNode *tail;
public:
    LRUCache(int capacity) : size(0), capacity(capacity) {
        // 虚拟头指针
        head = new ListNode(0, 0);
        tail = new ListNode(0, 0);
        head->next = tail;
        tail->prev = head;
    }

    int get(int key) {
        if (hashmap.find(key) == hashmap.end())
            return -1;
        auto node = hashmap.find(key)->second;
        // 双向链表移除节点
        auto prev_node = node->prev;
        auto next_node = node->next;
        prev_node->next = next_node;
        next_node->prev = prev_node;
        // 移动至头部
        head->next->prev = node;
        node->next = head->next;
        node->prev = head;
        head->next = node;
        return node->val;
    }

    void put(int key, int value) {
        // 存在key则update
        if (hashmap.find(key) != hashmap.end()) {
            auto node = hashmap.find(key)->second;
            node->val = value;
            this->get(key); // 刷新至最近使用
            return;
        }
        // 是否超出容量限制 淘汰一个缓存
        if (this->size == this->capacity) {
            auto del_node = tail->prev;
            del_node->next = nullptr;
            del_node->prev->next = tail;
            tail->prev = del_node->prev;
            del_node->prev = nullptr;
            hashmap.erase(del_node->key);
            delete del_node;
            --this->size;
        }
        // put
        auto put_node = new ListNode(key, value);
        hashmap[key] = put_node;
        put_node->next = head->next;
        head->next->prev = put_node;
        head->next = put_node;
        put_node->prev = head;
        ++this->size;
    }
};

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    auto *obj = new LRUCache(1);
    cout << obj->get(1) << "\n";    // -1
    obj->put(1, 1);
    cout << obj->get(1) << "\n";    // 1
    obj->put(1, 2);
    cout << obj->get(1) << "\n";    // 2
    obj->put(2, 3);
    cout << obj->get(1) << "\n";    // -1
    cout << obj->get(2) << "\n";    // 3
    return 0;
}

